### PR TITLE
ENT-3398: Upgrade PostgreSQL from 9.6.6 to 10.2

### PR DESCRIPTION
**Pull request by Lex-2008**
Ready for review!

Note that this is _preliminary_ result, for beta release. Follow-up to clean it up is created at [ENT-3648](https://tracker.mender.io/browse/ENT-3648)

We would like our beta testers to check new DB upgrade procedure using `pg_upgrade --link` and report if something went wrong.

### CODE REVIEW

File: packaging/common/cfengine-hub/postinstall.sh [hunk 1]
```diff
@@ -383,20 +397,20 @@ else
...
+  #     if ! [ -f "$PREFIX/state/pg/db_dump-$db.sql.gz" ]; then
+  #       continue
+  #     fi
+  #     cf_console echo "Restoring database $db..."
+  #     (cd /tmp && su cfpostgres -c "gunzip -c $PREFIX/state/pg/db_dump-$db.sql.gz | $PREFIX/bin/psql $db")
+  #     if [ $? != 0 ]; then
+  #       cf_console echo "Not able to migrate database $db. Backup is in $PREFIX/state/pg/db_dump-$db.sql.gz"
+  #     fi
+  #   done
+  # fi
```
Comment (kacf) [inline packaging/common/cfengine-hub/postinstall.sh]: Just remove this. We have the Git history if we need it again.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 2]
```diff
@@ -345,6 +345,20 @@ if [ ! -d $PREFIX/state/pg/data ]; then
   fi
 fi
 
+if is_upgrade && [ -d "$PREFIX/bak" ]; then
```
Comment (kacf) [inline packaging/common/cfengine-hub/postinstall.sh]: I would use a more descriptive name than `bak`.

File: packaging/common/cfengine-hub/preinstall.sh [hunk 3]
```diff
@@ -5,45 +5,45 @@ if is_upgrade; then
...
+#   done
+#
+#   if [ "$FAILED" != 0 ]; then
+#     for db in $CF_DBS; do
+#       rm -f "$PREFIX/state/pg/db_dump-$db.sql.gz"
+#     done
+#     exit 1
+#   fi
+#
+#   cf_console echo "Done making backup. They will be in $PREFIX/state/pg/db_dump-*.sql.gz."
```
Comment (kacf) [inline packaging/common/cfengine-hub/preinstall.sh]: Same here, remove.

**Comment by Lex-2008** [comment]
Amended according to @kacf 's suggestions: backup dir renamed to `$PREFIX/backup-before-postgres10-migration` (longer name to make it more noticeable)

File: deps-packaging/postgresql-hub/distfiles [hunk 4]
```diff
@@ -1 +1 @@
-5aebb23e96ebad0cc4d88aa4cc2b3c23  postgresql-9.6.6.tar.gz
+341d4f5a8eaced6df1a6e6a552c8a0b1  postgresql-10.2.tar.gz
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply deps-packaging/postgresql-hub/distfiles]: AFAIK, yes, we do

File: packaging/common/cfengine-hub/postinstall.sh [hunk 5]
```diff
@@ -345,6 +345,20 @@ if [ ! -d $PREFIX/state/pg/data ]; then
   fi
 fi
 
+BACKUP_DIR=$PREFIX/backup-before-postgres10-migration
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: It's a temporary dir which is removed automatically after successful migration. If migration wasn't successful - user will have to clean it up manually, and I'm not sure if having version in backup dir is helpful. Having UPGRADED_FROM.txt file, is, however.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 6]
```diff
@@ -345,6 +345,20 @@ if [ ! -d $PREFIX/state/pg/data ]; then
...
+if is_upgrade && [ -d "$BACKUP_DIR" ]; then
+  cf_console echo "Migrating database..."
+  (cd /tmp && cf_console time su cfpostgres -c "$PREFIX/bin/pg_upgrade --old-bindir=$BACKUP_DIR/bin --new-bindir=$PREFIX/bin --old-datadir=$BACKUP_DIR/data --new-datadir=$PREFIX/state/pg/data --link")
+  if [ $? = 0 ]; then
+    cf_console echo "Migration done, cleaning up"
+    rm -rf $BACKUP_DIR
+  else
+    cf_console echo "Migration failure. Manual interference required."
+    cf_console echo "Old stuff is saved in $BACKUP_DIR. Good luck!"
```
[the remark this answers is not in the document]
Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: And we'll release on May 4th?

[the remark this answers is not in the document]
Comment (nickanderson) [reply packaging/common/cfengine-hub/postinstall.sh]: ha if things aligned perfectly that would be haliarious

**Pull request merged**